### PR TITLE
[Feature] Support some downstream classification datasets.

### PR DESCRIPTION
**Pull request by zzc98**
## Motivation
Add some classification datasets. These datasets are listed below.

|  dataset  |  paper  | classes | size(train/test) |
|  :--:  | :--:  |  :--:  |  :--:  |
| [Oxford 102 Flowers](https://www.robots.ox.ac.uk/~vgg/data/flowers/102/) | [Automated flower classification over a large number of classes](https://www.robots.ox.ac.uk/~vgg/publications/2008/Nilsback08/nilsback08.pdf) | 102 | 2,040/6,149 |
| [Caltech-101](https://data.caltech.edu/records/mzrjq-6wc02) | [Learning Generative Visual Models from Few Training Examples: An Incremental Bayesian Approach Tested on 101 Object Categories](https://ieeexplore.ieee.org/document/1384978) | 102 | 3,060/6,084 |
| [Oxford-IIIT Pets](https://www.robots.ox.ac.uk/~vgg/data/pets/) | [0/1 Deep Neural Networks via Block Coordinate Descent ](https://arxiv.org/pdf/2206.09379v1.pdf) | 37 | 3,680/3,369 |
| [Describable Textures (DTD)](https://www.robots.ox.ac.uk/~vgg/data/dtd/) | [Describing Textures in the Wild](https://arxiv.org/pdf/1311.3618v2.pdf) | 47 | 3,760/1,880 |
| [FGVC Aircraft](https://www.robots.ox.ac.uk/~vgg/data/fgvc-aircraft/) |  [Fine-Grained Visual Classification of Aircraft](https://arxiv.org/pdf/1306.5151v1.pdf)| 100 | 6,667/3,333 |
| [Stanford Cars](https://ai.stanford.edu/~jkrause/cars/car_dataset.html) | [3D Object Representations for Fine-Grained Categorization](https://ieeexplore.ieee.org/document/6755945) | 196 | 8,144/8,041 |
| [SUN397](https://vision.princeton.edu/projects/2010/SUN/) | [SUN Database: Large-scale Scene Recognition from Abbey to Zoo](https://vision.princeton.edu/projects/2010/SUN/paper.pdf) | 397 | 19,850/19,850 |
| [Food-101](https://data.vision.ee.ethz.ch/cvl/datasets_extra/food-101/) | [Food-101 – Mining Discriminative Components with Random Forests](https://link.springer.com/chapter/10.1007/978-3-319-10599-4_29) | 101 | 75,750/25,250 |

## Examples

<details>
  <summary>Oxford 102 Flowers</summary>
  <pre><code>>>> from mmpretrain.datasets import Flowers102
>>> train_cfg = dict(data_root='data/Flowers102', split='trainval')
>>> train = Flowers102(**train_cfg)
>>> train
Dataset Flowers102
    Number of samples:  2040
    Root of dataset:    data/Flowers102
>>> test_cfg = dict(data_root='data/Flowers102', split='test')
>>> test = Flowers102(**test_cfg)
>>> test
Dataset Flowers102
    Number of samples:  6149
    Root of dataset:    data/Flowers102
</code></pre>
</details>

<details>
  <summary>Caltech-101</summary>
  <pre><code>>>> from mmpretrain.datasets import Caltech101
>>> train_cfg = dict(data_root='data/Caltech', split='train')
>>> train = Caltech101(**train_cfg)
>>> train
Dataset Caltech101
    Number of samples:  3060
    Number of categories:       102
    Root of dataset:    data/Caltech
>>> test_cfg = dict(data_root='data/Caltech', split='test')
>>> test = Caltech101(**test_cfg)
>>> test
Dataset Caltech101
    Number of samples:  6728
    Number of categories:       102
    Root of dataset:    data/Caltech
</code></pre>
</details>

<details>
  <summary>Oxford-IIIT Pets</summary>
  <pre><code>>>> from mmpretrain.datasets import OxfordIIITPet
>>> train_cfg = dict(data_root='data/Oxford-IIIT_Pets', split='trainval')
>>> train = OxfordIIITPet(**train_cfg)
>>> train
Dataset OxfordIIITPet
    Number of samples:  3680
    Number of categories:       37
    Root of dataset:    data/Oxford-IIIT_Pets
>>> test_cfg = dict(data_root='data/Oxford-IIIT_Pets', split='test')
>>> test = OxfordIIITPet(**test_cfg)
>>> test
Dataset OxfordIIITPet
    Number of samples:  3669
    Number of categories:       37
    Root of dataset:    data/Oxford-IIIT_Pets
</code></pre>
</details>

<details>
  <summary>Describable Textures (DTD)</summary>
  <pre><code>>>> from mmpretrain.datasets import DTD
>>> train_cfg = dict(data_root='data/dtd', split='trainval')
>>> train = DTD(**train_cfg)
>>> train
Dataset DTD
    Number of samples:  3760
    Number of categories:       47
    Root of dataset:    data/dtd
>>> test_cfg = dict(data_root='data/dtd', split='test')
>>> test = DTD(**test_cfg)
>>> test
Dataset DTD
    Number of samples:  1880
    Number of categories:       47
    Root of dataset:    data/dtd
</code></pre>
</details>

<details>
  <summary>FGVC Aircraft</summary>
  <pre><code>>>> from mmpretrain.datasets import FGVCAircraft
>>> train_cfg = dict(data_root='data/fgvc-aircraft-2013b/data', split='trainval')
>>> train = FGVCAircraft(**train_cfg)
>>> train
Dataset FGVCAircraft
    Number of samples:  6667
    Number of categories:       100
    Root of dataset:    data/fgvc-aircraft-2013b/data
>>> test_cfg = dict(data_root='data/fgvc-aircraft-2013b/data', split='test')
>>> test = FGVCAircraft(**test_cfg)
>>> test
Dataset FGVCAircraft
    Number of samples:  3333
    Number of categories:       100
    Root of dataset:    data/fgvc-aircraft-2013b/data
</code></pre>
</details>

<details>
  <summary>Stanford Cars</summary>
  <pre><code>>>> from mmpretrain.datasets import StanfordCars
>>> train_cfg = dict(data_root='data/Stanford_Cars', split='train')
>>> train = StanfordCars(**train_cfg)
>>> train
Dataset StanfordCars
    Number of samples:  8144
    Number of categories:       196
    Root of dataset:    data/Stanford_Cars
>>> test_cfg = dict(data_root='data/Stanford_Cars', split='test')
>>> test = StanfordCars(**test_cfg)
>>> test
Dataset StanfordCars
    Number of samples:  8041
    Number of categories:       196
    Root of dataset:    data/Stanford_Cars
</code></pre>
</details>

<details>
  <summary>SUN397</summary>
  <pre><code>>>> from mmpretrain.datasets import SUN397
>>> train_cfg = dict(data_root='data/SUN397', split='train')
>>> train = SUN397(**train_cfg)
>>> train
Dataset SUN397
    Number of samples:  19824
    Number of categories:       397
    Root of dataset:    data/SUN397
>>> test_cfg = dict(data_root='data/SUN397', split='test')
>>> test = SUN397(**test_cfg)
>>> test
Dataset SUN397
    Number of samples:  19829
    Number of categories:       397
    Root of dataset:    data/SUN397
</code></pre>
</details>

<details>
  <summary>Food-101</summary>
  <pre><code>>>> from mmpretrain.datasets import Food101
>>> train_cfg = dict(data_root='data/food-101', split='train')
>>> train = Food101(**train_cfg)
>>> train
Dataset Food101
    Number of samples:  75750
    Number of categories:       101
    Root of dataset:    data/food-101
>>> test_cfg = dict(data_root='data/food-101', split='test')
>>> test = Food101(**test_cfg)
>>> test
Dataset Food101
    Number of samples:  25250
    Number of categories:       101
    Root of dataset:    data/food-101
</code></pre>
</details>

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by wangbo-zhao** [comment]
The number of categories in FGVC Aircraft should be 100

**Comment by zzc98** [comment]
> The number of categories in FGVC Aircraft should be 100

The number of categories has been modified. Thanks.

File: mmpretrain/datasets/oxford102flowers.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+import mat4py
+from mmengine import get_file_backend
+
+from mmpretrain.registry import DATASETS
+from .base_dataset import BaseDataset
+
+
+@DATASETS.register_module()
+class Oxford102Flowers(BaseDataset):
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/oxford102flowers.py:12]: Flowers102 exists, please rm this file

File: mmpretrain/datasets/oxfordiiitpet.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        test_mode (bool): ``test_mode=True`` means in test phase. It determines
+             to use the training set or test set.
+        ann_file (str, optional): Annotation file path, path relative to
+            ``data_root``. Defaults to 'annotations/trainval.txt'.
+        data_prefix (str): Prefix for images, path relative to
+            ``data_root``. Defaults to 'images'.
+
+    Examples:
+        >>> from mmpretrain.datasets import OxfordIIITPet
+        >>> pet_train_cfg = dict(data_root='data/Oxford-IIIT_Pets')
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/oxfordiiitpet.py:43]: ```dict(data_root='data/Oxford-IIIT_Pets', ann_file = 'annotations/trainval.txt', )```
Although the default value is given, it is best to add ```ann_file =xxxx```

File: mmpretrain/datasets/stanford_cars.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+@DATASETS.register_module()
+class StanfordCars(BaseDataset):
+    """The Stanford Cars Dataset.
+
+    Support the `Stanford Cars Dataset <https://ai.stanford.edu/~jkrause/cars/car_dataset.html>`_ Dataset.
+    The official website provides two ways to organize the dataset.
+    Therefore, after downloading and decompression, the dataset directory structure is as follows.
+
+    Stanford Cars dataset directory: ::
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/stanford_cars.py:20]: why here are two ways?

File: mmpretrain/datasets/sun397.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        >>> test
+        Dataset SUN397
+            Number of samples:  19829
+            Number of categories:       397
+            Root of dataset:    data/SUN397
+    """  # noqa: E501
+
+    METAINFO = {'classes': SUN397_CATEGORIES}
+
+    INVALID = [
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/sun397.py:71]: put it out of the class, and add some comments

File: mmpretrain/datasets/dtd.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            Number of samples:  1880
+            Number of categories:       47
+            Root of dataset:    data/dtd
+    """  # noqa: E501
+
+    METAINFO = {'classes': DTD_CATEGORIES}
+
+    def __init__(self,
+                 data_root: str,
+                 mode: str = 'trainval',
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/dtd.py:69]: I think `split` is better than `mode`.

File: mmpretrain/datasets/categories.py [hunk 6]
```diff
@@ -1141,3 +1141,308 @@
...
+    '/l/lift_bridge', '/l/lighthouse', '/l/limousine_interior',
+    '/l/living_room', '/l/lobby', '/l/lock_chamber', '/l/locker_room',
+    '/m/mansion', '/m/manufactured_home', '/m/market/indoor',
+    '/m/market/outdoor', '/m/marsh', '/m/martial_arts_gym', '/m/mausoleum',
+    '/m/medina', '/m/moat/water', '/m/monastery/outdoor', '/m/mosque/indoor',
+    '/m/mosque/outdoor', '/m/motel', '/m/mountain', '/m/mountain_snowy',
+    '/m/movie_theater/indoor', '/m/museum/indoor', '/m/music_store',
+    '/m/music_studio', '/n/nuclear_power_plant/outdoor', '/n/nursery',
+    '/o/oast_house', '/o/observatory/outdoor', '/o/ocean', '/o/office',
+    '/o/office_building', '/o/oil_refinery/outdoor', '/o/oilrig',
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/categories.py:1365]: In any case, adding a first letter folder in front of the category name is very strange. It is best for the category name to be one word.

**Review by Ezra-Yu — COMMENTED** [review]
please rm the useless files.

File: docs/en/api/datasets.rst [hunk 7]
```diff
@@ -81,3 +81,44 @@ The dataset wrappers in the MMEngine can be directly used in MMPreTrain.
      - A wrapper of repeated dataset.
    * - :class:`~mmengine.dataset.ClassBalancedDataset`
      - A wrapper of class balanced dataset.
+
```
Comment (Ezra-Yu) [inline docs/en/api/datasets.rst:84]: put this section in right place.

File: mmpretrain/datasets/caltech101.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        >>> from mmpretrain.datasets import Caltech101
+        >>> train_cfg = dict(data_root='data/Caltech', split='train')
+        >>> train = Caltech101(**train_cfg)
+        >>> train
+        Dataset Caltech101
+            Number of samples:  3060
+            Number of categories:       102
+            Root of dataset:    data/Caltech
+        >>> test_cfg = dict(data_root='data/Caltech', split='test')
+        >>> test = Caltech101(**test_cfg)
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/caltech101.py:57]: ```test_dataset = Caltech101(data_root='data/Caltech', split='test')``` is enough. The current configuration items are relatively short, the lines about 'test_cfg', 'train_cfg' can be removed.

```
        >>> from mmpretrain.datasets import Caltech101
        >>> Caltech101(data_root='data/Caltech', split='train')  # train dataset
        Dataset Caltech101
               Number of samples:  3060
               Number of categories:       102
               Root of dataset:    data/Caltech
```

File: mmpretrain/datasets/cifar.py [hunk 9]
```diff
@@ -56,12 +55,30 @@ class CIFAR10(BaseDataset):
...
+
+        splits = ['train', 'test']
+        assert split in splits, \
+            f'Split {split} is not in default splits {splits}'
+        self.split = split
+
+        if split == 'train' and test_mode:
+            logger = MMLogger.get_current_instance()
+            logger.warning(
+                'split="train" but test_mode=True. The training set '
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/cifar.py:74]: 'split="train" but test_mode=True. The training set will be used.' is enough

File: mmpretrain/datasets/cifar.py [hunk 10]
```diff
@@ -56,12 +55,30 @@ class CIFAR10(BaseDataset):
...
+                'will be used. Please set split="test" (test) or '
+                'test_mode=False (train).')
+
+        if not data_root and not data_prefix:
+            raise RuntimeError('Please set ``data_root`` or ``data_prefix`` \
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/cifar.py:79]: 'Please set ``data_root``' user should only know ``data_root`` and ``split``.

File: mmpretrain/datasets/cifar.py [hunk 11]
```diff
@@ -56,12 +55,30 @@ class CIFAR10(BaseDataset):
...
+                 data_prefix: str = '',
+                 test_mode: bool = False,
                  **kwargs):
+
+        splits = ['train', 'test']
+        assert split in splits, \
+            f'Split {split} is not in default splits {splits}'
+        self.split = split
+
+        if split == 'train' and test_mode:
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/cifar.py:71]: add a comment `to handle the BC-breaking`.

File: mmpretrain/datasets/fgvcaircraft.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    METAINFO = {'classes': FGVCAIRCRAFT_CATEGORIES}
+
+    def __init__(self, data_root: str, split: str = 'trainval', **kwargs):
+
+        splits = ['train', 'val', 'trainval', 'test']
+        assert split in splits, \
+            f'Split {split} is not in default splits {splits}'
+        self.split = split
+
+        if split == 'train':
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/fgvcaircraft.py:64]: ```ann_file = f'images_variant_{split}.txt'```

File: mmpretrain/datasets/fgvcaircraft.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+@DATASETS.register_module()
+class FGVCAircraft(BaseDataset):
+    """The FGVC_Aircraft Dataset.
+
+    Support the `FGVC_Aircraft Dataset <https://www.robots.ox.ac.uk/~vgg/data/fgvc-aircraft/>`_ Dataset.
+    After downloading and decompression, the dataset directory structure is as follows.
+
+    FGVC Aircraft dataset directory: ::
+
+        fgvc-aircraft-2013b/data (data_root)/
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/fgvcaircraft.py:20]: ```fgvc-aircraft-2013b/data``` is set to data_root is not intuitive between other datasets.

File: mmpretrain/datasets/dtd.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            Root of dataset:    data/dtd
+    """  # noqa: E501
+
+    METAINFO = {'classes': DTD_CATEGORIES}
+
+    def __init__(self, data_root: str, split: str = 'trainval', **kwargs):
+
+        splits = ['train', 'val', 'trainval', 'test']
+        assert split in splits, \
+            f'Split {split} is not in default splits {splits}'
```
Comment (Ezra-Yu) [inline mmpretrain/datasets/dtd.py:67]: f"The `split` must be one of {splits}, but get '{split}'"

**Review by Ezra-Yu — COMMENTED** [review]
please recorrect this param doc. https://github.com/open-mmlab/mmpretrain/blob/e80418a424aaefb81c95df458216bb3e9af246c4/mmpretrain/datasets/base_dataset.py#L52

File: mmpretrain/datasets/flowers102.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+@DATASETS.register_module()
+class Flowers102(BaseDataset):
+    """The Oxford 102 Flower Dataset.
+
+    Support the `Oxford 102 Flowers Dataset <https://www.robots.ox.ac.uk/~vgg/data/flowers/102/>`_ Dataset.
+    After downloading and decompression, the dataset directory structure is as follows.
+
+    Flowers102 dataset directory: ::
+
+        Flowers102 (data_root)/
```
Comment (fangyixiao18) [inline mmpretrain/datasets/flowers102.py:20]: no need to write `(data_root)` here, as we give the sample to write the config and build the dataset below

Comment (fangyixiao18) [reply mmpretrain/datasets/flowers102.py:20]: please also check other docstring

File: mmpretrain/datasets/oxfordiiitpet.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+@DATASETS.register_module()
+class OxfordIIITPet(BaseDataset):
+    """The Oxford-IIIT Pets Dataset.
+
+    Support the `Oxford-IIIT Pets Dataset <https://www.robots.ox.ac.uk/~vgg/data/pets/>`_ Dataset.
+    After downloading and decompression, the dataset directory structure is as follows.
+
+    OxfordIIITPet_CATEGORIES dataset directory: ::
+
+        OxfordIIITPet_CATEGORIES (data_root)/
```
Comment (fangyixiao18) [inline mmpretrain/datasets/oxfordiiitpet.py:20]: The name here is not the same as it in example? Is it expected?
in example, `data_root='data/Oxford-IIIT_Pets`

File: mmpretrain/datasets/stanfordcars.py [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+        │   ├── 00001.jpg
+        │   ├── 00002.jpg
+        │   └── ...
+        └── cars_annos.mat
+
+    or
+
+    Stanford Cars dataset directory: ::
+
+        Stanford Cars (data_root)
```
Comment (fangyixiao18) [inline mmpretrain/datasets/stanfordcars.py:33]:
```suggestion
        Stanford_Cars
```
keep the same as it in the example

**Review by Ezra-Yu — APPROVED** [review]
LGTM.

**Review by fangyixiao18 — APPROVED** [review]

**Pull request merged**